kitsune/wiki/views.py

```diff
@@ -471,106 +471,144 @@ def edit_document(request, document_slug, revision_id=None):
-                    # Get the possibly new slug for the imminent redirection:
-                    try:
-                        doc = doc_form.save(None)
-                    except (TitleCollision, SlugCollision) as e:
-                        # TODO: .add_error() when we upgrade to Django 1.7
-                        errors = doc_form._errors.setdefault("title", ErrorList())
-                        message = "The {type} you selected is already in use."
-                        message = message.format(
-                            type="title" if isinstance(e, TitleCollision) else "slug"
-                        )
-                        errors.append(_(message))
-                    else:
-                        # Do we need to rebuild the KB?
-                        _maybe_schedule_rebuild(doc_form)
+        topics = []
+        for t in post_data.getlist("topics"):
+            topics.append(int(t))
+        post_data.setlist("topics", topics)
 
-                        return HttpResponseRedirect(
-                            urlparams(
-                                reverse("wiki.edit_document", args=[doc.slug]), opendescription=1
-                            )
-                        )
-                disclose_description = True
+        doc_form = DocumentForm(
+            post_data,
+            instance=doc,
+            can_archive=can_archive,
+            can_edit_needs_change=can_edit_needs_change,
+        )
+        if doc_form.is_valid():
+            # Get the possibly new slug for the imminent redirection:
+            try:
+                doc = doc_form.save(None)
+            except (TitleCollision, SlugCollision) as e:
+                # TODO: .add_error() when we upgrade to Django 1.7
+                errors = doc_form._errors.setdefault("title", ErrorList())
+                message = "The {type} you selected is already in use."
+                message = message.format(type="title" if isinstance(e, TitleCollision) else "slug")
+                errors.append(_(message))
             else:
-                raise PermissionDenied
-        elif which_form == "rev":
-            if doc.allows(user, "create_revision"):
-                rev_form = RevisionForm(request.POST)
-                rev_form.instance.document = doc  # for rev_form.clean()
-                if rev_form.is_valid():
-                    _save_rev_and_notify(rev_form, user, doc, base_rev=rev)
-                    if "notify-future-changes" in request.POST:
-                        EditDocumentEvent.notify(request.user, doc)
+                # Do we need to rebuild the KB?
+                _maybe_schedule_rebuild(doc_form)
 
-                    return HttpResponseRedirect(
-                        reverse("wiki.document_revisions", args=[document_slug])
-                    )
-            else:
-                raise PermissionDenied
+                return HttpResponseRedirect(urlparams(reverse("wiki.document", args=[doc.slug])))
```

Comment:
If the user changes the slug, the current document (with the original slug) becomes a redirect to the document at the new slug, so in that case, this code would redirect the user to original slug first, which would, in turn, then redirect to the new slug. I suspect this leaves users confused about what happened to the document at the original slug, so maybe we should add `?redirect="no"` so that the user is shown the original document. That way, users see what happened to the document they just edited, which has been turned into a redirect with a link to the new document slug in the body, and also clearly see that they can simply click that link if they'd like to go to the new document slug. I don't know. Maybe it's best to ask the content team what they'd prefer?
```suggestion
                return HttpResponseRedirect(urlparams(reverse("wiki.document", args=[doc.slug]), redirect="no"))
```

Reply:
By the way, if we decide not to do this, I just wanted to note that the `urlparams` call is not needed and can be removed.
```suggestion
                return HttpResponseRedirect(reverse("wiki.document", args=[doc.slug]))
```

Reply:
I hadn't thought about that at all! Good catch.